python/ray/data/__init__.py

```diff
@@ -30,6 +31,67 @@
 _cached_fn = None
 _cached_cls = None
 
+
+# Register custom Arrow array serializer to work around zero-copy slice pickling bug.
+# See https://issues.apache.org/jira/browse/ARROW-10739.
+def _register_arrow_array_serializer():
+    import ray
+    from ray.data.impl.arrow_block import _arrow_array_reduce
+    from ray.data.extensions import ArrowTensorArray
+
+    try:
+        import pyarrow as pa
+    except ModuleNotFoundError:
+        # No pyarrow installed so not using Arrow, so no need for custom serializer.
+        return
+
+    array_types = [
+        pa.lib.NullArray,
+        pa.lib.BooleanArray,
+        pa.lib.UInt8Array,
+        pa.lib.UInt16Array,
+        pa.lib.UInt32Array,
+        pa.lib.UInt64Array,
+        pa.lib.Int8Array,
+        pa.lib.Int16Array,
+        pa.lib.Int32Array,
+        pa.lib.Int64Array,
+        pa.lib.Date32Array,
+        pa.lib.Date64Array,
+        pa.lib.TimestampArray,
+        pa.lib.Time32Array,
+        pa.lib.Time64Array,
+        pa.lib.DurationArray,
+        pa.lib.MonthDayNanoIntervalArray,
+        pa.lib.HalfFloatArray,
+        pa.lib.FloatArray,
+        pa.lib.DoubleArray,
+        pa.lib.ListArray,
+        pa.lib.LargeListArray,
+        pa.lib.MapArray,
+        pa.lib.FixedSizeListArray,
+        pa.lib.UnionArray,
+        pa.lib.BinaryArray,
+        pa.lib.StringArray,
+        pa.lib.LargeBinaryArray,
+        pa.lib.LargeStringArray,
+        pa.lib.DictionaryArray,
+        pa.lib.FixedSizeBinaryArray,
+        pa.lib.Decimal128Array,
+        pa.lib.Decimal256Array,
+        pa.lib.StructArray,
+        pa.lib.ExtensionArray,
+        ArrowTensorArray,
+    ]
```

Comment:
This is defined within `pyarrow`, but is private behind their Cython layer: https://github.com/apache/arrow/blob/bdea1b298866be3efdfc295cf5b8afc68fb7fa61/python/pyarrow/array.pxi#L2538-L2575

Reply:
This seems to introduce a manual maintenance task going forward, since we cannot simply change pyarrow version number to upgrade. Is there a mechanism to track or automate this kind of task?

Reply:
Agreed! I haven't thought of a good way to automatically cover this, since if we could automatically generate the complete array types that we need to cover, we'd directly use such a list instead of defining our own. We should also keep in mind that this will be a stopgap fix and that we fully intend to fix this in upstream Arrow before the 8.0.0 release, at which point we'll most likely require that as the minimum Arrow version.

In the meantime, there are the following cases for future breaks:
1. Our custom serialization breaks for one of these array types. We can boost our test coverage to cover all major array types (currently working on this) so this is caught in our CI when a new pyarrow version is released.
2. Arrow adds a new array type that's not in this list. Note that this won't cause a failure; instead, this will cause the default serialization path to be used. Since these array types will most likely be very niche, I think that addressing these as they come up should be fine.
3. Arrow removes an array type that's in this list. This should be caught in CI via a very loud `AttributeError` that will fail every Datasets test, so this will be easy to catch.

The biggest pain around (1) and (2) is that pushing out a new Datasets patch release carries a lot of overhead, so it would be great if we could ensure that such breaks don't block users. Here are some ideas for mitigating this issue:
a. Put a try-except around the custom serialization code; upon failure, log the error and fall back to the default serialization path.
b. Put the registering of the custom serializer behind an environment variable flag that defaults to being enabled. This way, if the user runs into one of these breaking cases that isn't fix by the try-except fallback, they can always completely disable the custom serializer.
c. Add a `_register_serializer_for_array_type()` hook, so if Arrow adds a new array type and it works with the common `pa.Array` base class path of our custom serializer (which it most likely will), a user can add a `_register_serializer_for_array_type(pa.lib.SomeNewArray)` before their Datasets code to try to leverage the custom serialization path.

I think that decent test coverage and doing (a), (b), and (c) should mitigate the pain caused by any of these breaks. What do you think?

Reply:
Sounds good, before a root fix in upstream (nice that we already have this in plan), catching the errors and having a fallback to default serialization make a lot sense (eg. we will not crash, at worst being inefficient -- for unknown types, if any at all). Thanks for putting thought into it!